### PR TITLE
[CLOUDTRUST-2109] Authorization refactoring

### PR DESCRIPTION
**Pull request by harture**
We plan to split the authorizations refactoring in 2 phases (2 PR).
First, we deploy the 1st PR which contains the authorization edition/management. We perform the authorizations configuration which will be stored in DB. This config will not be used, the JSON is still used to check the authorizations.
Second step, we deploy the 2nd PR which will remove the usage of JSOn and compute the authorization from the DB configuration.
This strategy should allow us to minimize the downtime and migration risks.

### CODE REVIEW

**Comment by harture** [comment]
We plan to split the authorizations refactoring in 2 phases (2 PR).
First, we deploy the 1st PR which contains the authorization edition/management. We perform the authorizations configuration which will be stored in DB. This config will not be used, the JSON is still used to check the authorizations.
Second step, we deploy the 2nd PR which will remove the usage of JSOn and compute the authorization from the DB configuration.
This strategy should allow us to minimize the downtime and migration risks.

File: internal/keycloakb/configdbmodule.go [hunk 1]
```diff
@@ -67,3 +76,76 @@ func (c *configurationDBModule) GetConfiguration(context context.Context, realmI
 		return config, err
 	}
 }
+
+func (c *configurationDBModule) GetAuthorizations(ctx context.Context, realmID string, groupName string) ([]dto.Authorization, error) {
```
Comment (fperot74) [inline internal/keycloakb/configdbmodule.go]: May be reused in paper-card-service

File: internal/keycloakb/configdbmodule_test.go [hunk 2]
```diff
@@ -8,25 +8,29 @@ import (
...
+	"github.com/cloudtrust/common-service/log"
 	"github.com/cloudtrust/keycloak-bridge/internal/dto"
 	"github.com/cloudtrust/keycloak-bridge/internal/keycloakb/mock"
+	msg "github.com/cloudtrust/keycloak-bridge/internal/messages"
 	"github.com/golang/mock/gomock"
 	"github.com/stretchr/testify/assert"
 )
 
-func useMockDB(t *testing.T, fn func(DBConfiguration)) {
+func useMockDB(t *testing.T, fn func(sqltypes.CloudtrustDB)) {
```
Comment (fperot74) [inline internal/keycloakb/configdbmodule_test.go]: This function is not used... it can be removed

File: pkg/events/component.go [hunk 3]
```diff
@@ -40,6 +42,23 @@ func (ec *component) reportEvent(ctx context.Context, apiCall string, values ...
 
 }
 
+// Get actions
+func (ec *component) GetActions(ctx context.Context) ([]api.ActionRepresentation, error) {
```
Comment (fperot74) [inline pkg/events/component.go]: Could be part of the common lib

File: api/management/api.go [hunk 4]
```diff
@@ -242,6 +254,97 @@ func ConvertToKCUser(user UserRepresentation) kc.UserRepresentation {
 	return userRep
 }
 
+// ConvertToKCGroup creates a KC group representation from an API group
+func ConvertToKCGroup(group GroupRepresentation) kc.GroupRepresentation {
+	var groupRep kc.GroupRepresentation
+
+	groupRep.Name = group.Name
+
+	return groupRep
```
[the remark this answers is not in the document]
Comment (harture) [reply api/management/api.go]: Thanks, indeed it is better like that ! ;)

File: internal/keycloakb/configdbmodule.go (hunk 1 above)
Comment (harture) [reply internal/keycloakb/configdbmodule.go]: When we will retrieve the authorizations to apply authorizations, we will retrieve all of them not just one for a specific realm and group.

File: internal/keycloakb/configdbmodule_test.go (hunk 2 above)
Comment (harture) [reply internal/keycloakb/configdbmodule_test.go]: thanks

File: pkg/events/component.go (hunk 3 above)
Comment (harture) [reply pkg/events/component.go]: I agree this method is very similar for each Component but a simple and elegant way to do it ...
Would you have any idea ?

Comment (fperot74) [reply pkg/events/component.go]: Let's say we will have a look at this when we'll have time for this... (it almost means never)

File: api/events/api.go [hunk 5]
```diff
@@ -2,6 +2,12 @@ package events_api
 
 import "database/sql"
 
+// ActionRepresentation struct
+type ActionRepresentation struct {
```
Comment (bsoniam) [inline api/events/api.go]: The same structure is used by papercard module.  
If for the long term this stays the same, then we can move them in common service

File: api/events/swagger-api_events.yaml [hunk 6]
```diff
@@ -9,6 +9,20 @@ tags:
...
+      summary: Get the list of all possible actions
+      responses:
+        200:
+          description: successful operation
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  type: string
```
Comment (bsoniam) [inline api/events/swagger-api_events.yaml]: Is not the same as you code in the paper card?
  /cards/actions:
    get:
      tags:
      - Actions
      summary: Get the list of all possible actions
      responses:
        200:
          description: successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                 ** $ref: '#/components/schemas/Actions' **

File: pkg/statistics/authorization.go [hunk 7]
```diff
@@ -8,14 +8,27 @@ import (
 	api "github.com/cloudtrust/keycloak-bridge/api/statistics"
 )
 
+var actions []security.Action
+
+func newAction(as string, scope security.Scope) security.Action {
+	a := security.Action{
```
Comment (bsoniam) [inline pkg/statistics/authorization.go]: Same

File: api/events/swagger-api_events.yaml (hunk 6 above)
Comment (harture) [reply api/events/swagger-api_events.yaml]: You are right, it was a bug, it is fixed in the PR #159

File: pkg/events/authorization.go [hunk 8]
```diff
@@ -8,11 +8,24 @@ import (
 	api "github.com/cloudtrust/keycloak-bridge/api/events"
 )
 
+var actions []security.Action
+
+func newAction(as string, scope security.Scope) security.Action {
+	a := security.Action{
```
[the remark this answers is not in the document]
Comment (harture) [reply pkg/events/authorization.go]: I think I can remove this Id, we finally don't use it.

File: api/events/api.go (hunk 5 above)
Comment (harture) [reply api/events/api.go]: I will move it in common during PR2

File: api/management/swagger-api_management.yaml [hunk 9]
```diff
@@ -15,6 +15,20 @@ tags:
 - name: Roles
   description: Roles management
 paths:
+  /actions:
+    get:
+      tags:
```
[the remark this answers is not in the document]
Comment (harture) [reply api/management/swagger-api_management.yaml]: TODO in PR2, thks

File: internal/messages/errormessages.go [hunk 10]
```diff
@@ -33,8 +34,10 @@ const (
 	PhoneNumber        = "phoneNumber"
 	Gender             = "gender"
 	Birthdate          = "birthdate"
-	GroudID            = "groupId"
-	GroudIDs           = "groupIds"
+	GroupID            = "groupId"
+	GroupName          = "groupName"
+	Name               = "name"
```
[the remark this answers is not in the document]
Comment (harture) [reply internal/messages/errormessages.go]: I will fix it

**Pull request merged**